packages/block-library/src/image/edit.native.js

```diff
@@ -489,6 +540,9 @@ export class ImageEdit extends Component {
 				<PanelBody title={ __( 'Link Settings' ) }>
 					{ this.getLinkSettings( true ) }
 				</PanelBody>
+				{ androidOnly &&
+					typeof attributes.id !== 'undefined' &&
```

Comment:
Aha, I understand how checking for undefined `id` fixes [the crash](https://github.com/WordPress/gutenberg/pull/31705#issuecomment-841317965), but it's not clear why the code we are in here (the `getInspectorControls` function) should care about this condition which looks too implementation specific. Feels to me that's the responsibility of `getSetFeaturedButton` to care about, especially since the whole PanelBody is managed by `getSetFeaturedButton`.

What do you think about moving the "id undefined" condition inside `getSetFeaturedButton`? While at it, perhaps now the `androidOnly` condition could move there at the function top as well  now that there's going to be one there governing the whole visibility of the set/remove option.


There's one more point to raise here actually: is the undefined state of the `id` a valid state? If yes then indeed it makes sense to cover for it. But, if it's not, then it's a bug we need to fix elsewhere and not try to guard against it here since it will mostly "hide" the issue. What are your thoughts on this?

Reply:
The logic outlined in [the previous suggestion](https://github.com/WordPress/gutenberg/pull/31705#discussion_r632567432) to move the `androidOnly` conditional from `getSetFeaturedButton` to here is what led me to also add the check for `undefined` in the same place. Can you help me understand why this check would require a different treatment? The reasoning's not super clear to me just yet. 

> There's one more point to raise here actually: is the undefined state of the id a valid state? If yes then indeed it makes sense to cover for it. But, if it's not, then it's a bug we need to fix elsewhere and not try to guard against it here since it will mostly "hide" the issue. What are your thoughts on this?

`undefined` is a valid state, to my knowledge. An image only has an ID if it's been uploaded to the media library of the site it's being used with. There are valid use cases, such as with page templates, where images are uploaded to another site's media library.

Reply:
> The logic outlined in the previous suggestion to move the androidOnly conditional from getSetFeaturedButton to here is what led me to also add the check for undefined in the same place. Can you help me understand why this check would require a different treatment? The reasoning's not super clear to me just yet.

Oh sure, sorry. Lemme expand on how I perceive the two conditionals:
1. `androidOnly` is a conditional that can appear anywhere in the code, in an opportunistic way wherever we need to make something platform-specific. In this case, moving it inside `getInspectorControls()` helped un-indent and optimize some code while keeping the logic simple enough: Instead of just "get the set-featured-image button" it became "get the set-featured-image button on Android only".
2. `typeof attributes.id !== 'undefined'` on the other hand is very specific to the featured image logic and captures a very specific detail (don't try to offer the set/remove button if the image doesn't have an id) which the `getInspectorControls()`  shouldn't care about... this conditional is more of a responsibility of a function closer to the feature itself (`getSetFeaturedButton` in this case).

You don't _have_ to move any of conditionals if you don't quite see it fitting. In such a case, I'd probably suggest replacing `typeof attributes.id !== 'undefined'` with a function with a more descriptive name such as "canImageBeFeatured()" which sounds just enough for `getInspectorControls()` to care about, if that makes sense.

No strong feelings either way, let me know if my suggestions are more clear at this stage, thanks!

Reply:
> undefined is a valid state, to my knowledge. An image only has an ID if it's been uploaded to the media library of the site it's being used with. There are valid use cases, such as with page templates, where images are uploaded to another site's media library. Do you have further/different thoughts around that?

Perfect, thanks for the context and insight! Makes sense to keep treating a missing `id` as a valid value 👍. Best if we can capture this fact in a comment in the code (perhaps inside a "canImageBeFeatured()" function as I mentioned in the previous comment) as it's good context to have handy. WDYT?

Reply:
Great, that's clear, thank you so much for the extended explanation. 🙇‍♀️ I'll work on that!

Reply:
I've gone ahead to implement a `canImageBeFeatured` boolean and took the opportunity to use it with the `isFeaturedimage` boolean, as there was some code in it that served the same purpose. Does that look okay to you?

Reply:
Looks cool to me, thanks! I've only added a comment about adding a line comment or similar as there's quite some context packed into that check for `id` and it'd be best if we leave some context near it.